### PR TITLE
[ldap-groups] manage Rover groups based on AWS groups

### PR DESCRIPTION
**Pull request by chassing**
Adapt the `ldap-groups` integration to manage Rover groups based on AWS groups used for AWS SSO.

Ticket: [SDE-3516](https://issues.redhat.com/browse/SDE-3516)
Design doc: https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/94564
Depends on: https://github.com/app-sre/qontract-schemas/pull/584

### CODE REVIEW

**Review by fishi0x01 — APPROVED** [review]
:rocket:

File: reconcile/ldap_groups/integration.py [hunk 1]
```diff
@@ -129,14 +149,32 @@ def get_integration_settings(query_func: Callable) -> LdapGroupsSettingsV1:
...
-        duplicates = find_duplicates(role.ldap_group for role in roles)
-        if duplicates:
+        if duplicates := find_duplicates(role.ldap_group for role in roles):
             for dup in duplicates:
                 logging.error(f"{dup} is already in use by another role.")
             raise ValueError("Duplicate ldapGroup value(s) found.")
         return roles
 
-    def fetch_desired_state(
+    def get_aws_groups(self, query_func: Callable) -> list[AWSGroupV1]:
```
Comment (rporres) [inline reconcile/ldap_groups/integration.py:158]: Just your regular request of docstrings for public methods :)

**Review by rporres — APPROVED** [review]
documentation needs aside, this looks great.

File: reconcile/ldap_groups/integration.py [hunk 2]
```diff
@@ -129,14 +149,32 @@ def get_integration_settings(query_func: Callable) -> LdapGroupsSettingsV1:
...
+        data = aws_groups_query(query_func, variables={})
+        aws_groups = [
+            group
+            for group in data.aws_groups or []
+            if group.roles
+            and (group.account.sso is None or group.account.sso)
```
Comment (hemslo) [inline reconcile/ldap_groups/integration.py:164]: `sso` default to True if not specified? Seems a dangerous default.

File: reconcile/ldap_groups/integration.py [hunk 3]
```diff
@@ -129,14 +149,32 @@ def get_integration_settings(query_func: Callable) -> LdapGroupsSettingsV1:
...
+            # exclude roles without users
+            and [user for role in group.roles for user in role.users]
```
Comment (hemslo) [inline reconcile/ldap_groups/integration.py:166]: maybe use `any` to make it easier to understand

```python
any(role.users for role in group.roles)
```

File: reconcile/ldap_groups/integration.py [hunk 4]
```diff
@@ -157,6 +195,42 @@ def fetch_desired_state(
             if role.ldap_group
         ]
 
+    def get_desired_groups_for_aws_groups(
+        self,
+        aws_groups: Iterable[AWSGroupV1],
+        owners: Iterable[Entity],
```
Comment (hemslo) [inline reconcile/ldap_groups/integration.py:201]: `owners` is assigned to `Group.owners` directly, need it to be `list[Entity]`

```
Expected type 'list[Entity]', got 'Iterable[Entity]' instead 
```

File: reconcile/ldap_groups/integration.py (hunk 2 above)
Comment (chassing) [reply reconcile/ldap_groups/integration.py:164]: You're right! Let me change it!

File: reconcile/ldap_groups/integration.py (hunk 1 above)
Comment (chassing) [reply reconcile/ldap_groups/integration.py:164]: Added docs string to all public methods.

**Review by hemslo — APPROVED** [review]
LGTM!

File: reconcile/ldap_groups/integration.py (hunk 1 above)
Comment (rporres) [reply reconcile/ldap_groups/integration.py:164]: you rock! <3

**Pull request merged**